server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -225,8 +233,15 @@ private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T>
                 final LeafReaderContext[] leaves = leafSlices[i].leaves;
                 final C collector = collectors.get(i);
                 FutureTask<C> task = new FutureTask<>(() -> {
-                    search(Arrays.asList(leaves), weight, collector);
-                    return collector;
+                    try {
+                        search(Arrays.asList(leaves), weight, collector);
+                        return collector;
+                    } finally {
+                        if (collector instanceof TwoPhaseCollector twoPhaseCollector) {
+                            twoPhaseCollector.postCollection();
```

Comment:
This change change is the reason of this PR. We check whether collector is a `TwoPhaseCollector ` and then run post collect. This makes sense. 

I'm wondering whether can reduce the number of places elsewhere where we use `TwoPhaseCollector` as type? To essentially reduce the surface area of this change.